src/vulkan/Adapter.zig

```diff
@@ -0,0 +1,222 @@
+        features.independent_blend == vk.TRUE and
+        features.sample_rate_shading == vk.TRUE and
+        // At least one of the following texture compression forms
+        (features.texture_compression_bc == vk.TRUE or
+        features.texture_compression_etc2 == vk.TRUE or
+        features.texture_compression_astc_ldr == vk.TRUE);
+}
+
+fn rateDevice(
+    props: vk.PhysicalDeviceProperties,
+    features: vk.PhysicalDeviceFeatures,
+    power_preference: gpu.PowerPreference,
+) u32 {
+    // Can't function without geometry shaders
+    if (features.geometry_shader == vk.FALSE) {
+        return 0;
+    }
+
+    var score: u32 = 0;
+    switch (props.device_type) {
+        .integrated_gpu => if (power_preference == .low_power) {
+            score += 1000;
+        },
+        .discrete_gpu => if (power_preference == .high_performance) {
+            score += 1000;
+        },
+        else => {},
+    }
+
+    score += props.limits.max_image_dimension_2d;
+
+    return score;
+}
+
+const QueueFamilies = struct {
+    graphics: u32,
+    compute: u32,
+};
+
+fn findQueueFamilies(instance: Instance, device: vk.PhysicalDevice) !?QueueFamilies {
+    var queue_family_count: u32 = 0;
+    _ = instance.dispatch.getPhysicalDeviceQueueFamilyProperties(device, &queue_family_count, null);
+
+    var queue_families = try instance.allocator.alloc(vk.QueueFamilyProperties, queue_family_count);
+    defer instance.allocator.free(queue_families);
+    _ = instance.dispatch.getPhysicalDeviceQueueFamilyProperties(device, &queue_family_count, queue_families.ptr);
+
+    var graphics: ?u32 = null;
+    var compute: ?u32 = null;
+
+    for (queue_families, 0..) |family, i| {
+        if (family.queue_flags.graphics_bit and family.queue_flags.compute_bit) {
+            graphics = @intCast(i);
+            compute = @intCast(i);
+            break;
+        }
+
+        if (family.queue_flags.graphics_bit) {
+            graphics = @intCast(i);
+        }
+
+        if (family.queue_flags.compute_bit) {
+            compute = @intCast(i);
+        }
+
+        if (graphics != null and compute != null) {
+            break;
+        }
+    } else {
+        return null;
+    }
+
+    return .{
+        .graphics = graphics.?,
+        .compute = compute.?,
+    };
+}
+
+pub fn hasExtension(adapter: Adapter, name: []const u8) bool {
+    for (adapter.extensions) |ext| {
+        if (std.mem.eql(u8, name, std.mem.sliceTo(&ext.extension_name, 0))) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+pub fn createDevice(adapter: Adapter, descriptor: *const gpu.Device.Descriptor) !Device {
+    return Device.init(adapter, descriptor);
+}
+
+const VendorID = enum(u32) {
+    AMD = 0x1002,
+    Apple = 0x106b,
+    ARM = 0x13B5,
+    Google = 0x1AE0,
+    @"Img Tec" = 0x1010,
+    Intel = 0x8086,
+    Mesa = 0x10005,
+    Microsoft = 0x1414,
+    Nvidia = 0x10DE,
+    Qualcomm = 0x5143,
+    Samsung = 0x144d,
+    _,
```

Comment:
This enum should probably be made into more idiomatic names

Reply:
they are taken from https://github.com/hexops/dawn/blob/33aad65c769086285f0be4beb6036665bd4d8416/src/dawn/gpu_info.json#L78

but i see it makes sense because by mapping values to names, we can get rid of [this](https://github.com/hexops/mach-dusk/pull/30/files#diff-cc106d39cbc071e5a3ab48790db7deda5aae0ecdc2c6a572d691f8517ece38f9R99) :+1: